pageserver/src/tenant/delete.rs

```diff
@@ -538,16 +541,72 @@ impl DeleteTenantFlow {
             .context("cleanup_remaining_fs_traces")?;
 
         {
-            let mut locked = tenants.write().unwrap();
-            if locked.remove(&tenant.tenant_id).is_none() {
-                warn!("Tenant got removed from tenants map during deletion");
-            };
-
-            // FIXME: we should not be modifying this from outside of mgr.rs.
-            // This will go away when we simplify deletion (https://github.com/neondatabase/neon/issues/5080)
-            crate::metrics::TENANT_MANAGER
-                .tenant_slots
-                .set(locked.len() as u64);
+            pausable_failpoint!("tenant-delete-before-map-remove");
+
+            // This block is simply removing the TenantSlot for this tenant.  It requires a loop because
+            // we might conflict with a TenantSlot::InProgress marker and need to wait for it.
+            //
+            // This complexity will go away when we simplify how deletion works:
+            // https://github.com/neondatabase/neon/issues/5080
+            loop {
+                // Under the TenantMap lock, try to remove the tenant.  We usually succeed, but if
+                // we encounter an InProgress marker, yield the barrier it contains and wait on it.
+                let barrier = {
+                    let mut locked = tenants.write().unwrap();
+                    let removed = locked.remove(&tenant.tenant_id);
+
+                    // FIXME: we should not be modifying this from outside of mgr.rs.
+                    // This will go away when we simplify deletion (https://github.com/neondatabase/neon/issues/5080)
+                    crate::metrics::TENANT_MANAGER
+                        .tenant_slots
+                        .set(locked.len() as u64);
+
+                    match removed {
+                        TenantsMapRemoveResult::Occupied(slot) => {
+                            match slot {
+                                TenantSlot::Secondary => {
+                                    // This is unexpected: this secondary tenants should not have been created, and we
+                                    // are not in a position to shut it down from here.
+                                    tracing::warn!(
+                                        "Tenant transitioned to secondary mode while deleting!"
+                                    );
+                                }
+                                TenantSlot::Attached(tenant) => match tenant.current_state() {
+                                    TenantState::Stopping { .. } | TenantState::Broken { .. } => {
+                                        // Expected: we put the tenant into stopping state before we start deleting it
+                                    }
```

Comment:
I am under the impression that deletion does not touch broken, but it can make a tenant broken (like seen with #5917), but I think it would had happened earlier already?

Unsure if it matters here.. Looking around I am unsure where do we even restore an inprogress deletion, or is the TenantSlot::InProgress not being used for deletions? I vaguely remember that it is.

Reply:
The tolerance of Broken is meant to match the check done in DeleteTenantFlow::prepare -- it seems like the entire deletion operation can complete while the tenant is in a broken state, so we need to tolerate it here too